src/adapter/src/coord/timestamp_oracle.rs

```diff
@@ -65,6 +67,40 @@ pub trait TimestampOracle<T> {
     /// - r_0 <= input <= w_0 -> r_1 = input and w_1 = w_0
     /// - r_0 <= w_0 <= input -> r_1 = input and w_1 = input
     async fn apply_write(&mut self, write_ts: T);
+
+    /// Get a shared, shallow clone of the oracle. Returns `None` if this oracle
+    /// is not shareable.
+    fn get_shared(&self) -> Option<Arc<dyn ShareableTimestampOracle<T> + Send + Sync>>;
```

Comment:
There's two reasons:
 - `TimestampOracle` has some methods that take a `&mut` self, because that's what the old `CatalogOracle` needs. Because it does actually mutate itself. The `ShareableTimestampOracle` methods all only take `&self`, which is important because you can only call methods that take `&self` behind a plain `Arc<T>`. For the crdb oracle this works out because all the calls go to crdb and don't need to mutate oracle state.
 - `TimestampOracle` is `#[async_trait(?Send)]`, again because `CatalogOracle` cannot be `Send` (or `Sync`). So it cannot be send over to those async tasks that call `read_ts` asynchronously (off the coord main loop). `SharedTimestampOracle` is `#[async_trait]`, without the `?Send`.

And yes, subtle and annoying! 🙈